src/basic/recurse-dir.c

```diff
@@ -283,7 +290,7 @@ int recurse_dir(
                                         inode_fd = safe_close(inode_fd);
                                 }
 
-                        } else if (statx_mask != 0 || (de->entries[i]->d_type == DT_UNKNOWN && (flags & RECURSE_DIR_ENSURE_TYPE))) {
+                        } else if (statx_mask != 0) {
```

Comment:
the reason we did this logic here so far is to avoid two statx(), i.e. one in readdir_all(), and once here in case people set statx_mask != 0.

Hence, please leave this in place here, and change the readdir_all() call above to mask out RECURSE_DIR_ENSURE_TYPE.

That way RECURSE_DIR_ENSURE_TYPE can be passed to both readdir_all() and to recurse_dir(), and each times it will do the right thing, but we still won#t call statx() twice for each dirent in recurse_dir().